example/bwmf/bwmf_task.go

```diff
@@ -168,19 +288,68 @@ func (t *bwmfTask) doDataReady(ctx context.Context, fromID uint64, method string
 
 // These two function carry out actual optimization.
 func (t *bwmfTask) updateDShard(ctx context.Context) {
-	select {
-	case <-time.After(time.Duration(rand.Intn(3)) * time.Second):
-	case <-ctx.Done():
-		return
+
+	if t.dims.N == -1 {
+		t.dims.N = 0
+		for _, m := range t.peerShards {
+			t.dims.N += len(m.Row)
+		}
+		wh := make([][]float32, t.dims.N)
+		for i, _ := range wh {
+			wh[i] = make([]float32, t.dims.m)
+		}
+		t.dLoss = &KLDivLoss{
+			V:  t.rowShard,
+			WH: wh,
+			m:  t.dims.N,
+			n:  t.dims.m,
+			k:  t.dims.k,
+		}
+	}
+
+	t.dLoss.W = NewBlocksParameter(&t.peerShards)
+	param := NewSingleBlockParameter(t.dShard)
+
+	ok := t.optimizer.Minimize(t.dLoss, t.stopCriteria, param)
```

Comment:
Where?
If you mean
```
select{
case t.updateDone<-
case <-ctx.Done()
```

This is totally wrong. I meant to check ctx.Done with Minimize. Just think of Minimize returns a future. And you can do future.Cancel() because you don't need that operation anymore.
Or maybe you still want to finish the update before starting another update. I'm not sure.

Anyway, I'm thinking about the case of task failure, which I think complicates our process. Let's ignore any failure case now.